File: Google/729. My Calendar I.js

var MyCalendar = function() {
	this.bookings = null;
};

/** 
 * @param {number} start 
 * @param {number} end
 * @return {boolean}
 */
MyCalendar.prototype.book = function(start, end) {
	if (!this.bookings) {
		this.bookings = new TreeNode(start, end);
		return true;
	}

	return insert(this.bookings, start, end);
};

var insert = function(node, start, end) {
	if (end <= node.start) {
		if (!node.left) {
			node.left = new TreeNode(start, end);
			return true;
		} else return insert(node.left, start, end);
	} else if (start >= node.end) {
		if (!node.right) {
			node.right = new TreeNode(start, end);
			return true;
		} else return insert(node.right, start, end);
	} else return false;
}

class TreeNode {
	constructor(start, end) {
		this.start = start;
		this.end = end;
		this.left = null;
		this.right = null;
	}
};

/** 
 * Your MyCalendar object will be instantiated and called as such:
 * var obj = new MyCalendar()
 * var param_1 = obj.book(start,end)
 */

 const calendar = new MyCalendar();
 calendar.book(47, 50);
 calendar.book(33, 41);
 calendar.book(39, 45);

File: Google/729. My Calendar I.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace LeetCode
{
    public class Program
    {
        static void Main(string[] args)
        {
            var calendar = new MyCalendar();
            calendar.Book(10, 20);
            calendar.Book(15, 25);
            calendar.Book(20, 30);
        }
    }

    public class MyCalendar
    {
        private SortedList<int, int> Bookings { get; set; }

        public MyCalendar()
        {
            this.Bookings = new SortedList<int, int>();
        }

        public bool Book(int start, int end)
        {
            var ends = this.Bookings.Keys;
            var starts = this.Bookings.Values;
            var size = ends.Count;

            var pos = -1;
            for (var i = 0; i < size; i++)
            {
                if (start < ends[i]) break;
                
                pos = i;
            }

            if (size == 0 ||
                pos < 0 && end <= starts[0] ||
                pos == ends.Count - 1 ||
                end <= starts[pos + 1])
            {
                this.Bookings.Add(end, start);
                return true;
            }

            return false;
        }
    }

    /**
     * Your MyCalendar object will be instantiated and called as such:
     * MyCalendar obj = new MyCalendar();
     * bool param_1 = obj.Book(start,end);
     */
}
